File: BasicTypes/src/main/java/org/eso/ias/types/AlarmState.java
package org.eso.ias.types;

/** 
 * The state and state machine of an Alarm
 */
public enum AlarmState {

    /** Alarm clear and acknowledged */
	CLEAR_ACK,

	/** Alarm clear and unacknowledged */
	CLEAR_UNACK,

	/** Alarm set and unacknowledged */
	SET_UNACK,

	/** Alarm set and acknowledged */
	SET_ACK;
	
	/**
	 * 
	 * @return <code>true</code> if the alarm is set;
	 *         <code>false</code> otherwise
	 */
	public final boolean isSet() {
		return this==SET_ACK || this==SET_UNACK;
	}

	/**
	 *
	 * @return true if the alarm has been acknowledged
	 * 		   false otherwise
	 */
	public final boolean isAcked() {
		return this==SET_ACK || this==CLEAR_ACK;
	}

    /** The alarm has been Acked (by operator) */
    public AlarmState ack() {
		switch (this) {
			case SET_ACK:
			case CLEAR_ACK: return this;
			case SET_UNACK: return SET_ACK;
			case CLEAR_UNACK: return CLEAR_ACK;
			default: throw new UnsupportedOperationException("Unknown alarm state "+this);
		}
	}

    /** The alarm is set (by the TF) */
    public AlarmState set() {
		switch (this) {
			case SET_ACK:
			case SET_UNACK: return this;
			case CLEAR_ACK:
			case CLEAR_UNACK: return SET_UNACK;
			default: throw new UnsupportedOperationException("Unknown alarm state "+this);
		}
	}

    /** The alarm is clear (by the TF) */
    public AlarmState clear() {
		switch (this) {
			case CLEAR_ACK:
			case CLEAR_UNACK: return this;
			case SET_ACK: return CLEAR_ACK;
			case SET_UNACK: return CLEAR_UNACK;
			default: throw new UnsupportedOperationException("Unknown alarm state "+this);
		}
	}
}

File: BasicTypes/src/main/python/IasBasicTypes/AlarmState.py
'''
The alarm state and state machine of alarms as defined in org.eso.ias.types.AlarmState
'''

from enum import Enum

class AlarmState(Enum):

    # Alarm clear and acknowledged
    CLEAR_ACK = 0

    # Alarm clear and unacknowledged
    CLEAR_UNACK = 1

    # Alarm set and unacknowledged
    SET_UNACK = 2

    # Alarm set and acknowledged
    SET_ACK = 3

    def is_set(self):
        '''
        :return True if the alarm is set; False otherwise
        '''
        return self==AlarmState.SET_ACK or self==AlarmState.SET_UNACK

    def is_acked(self):
        '''
        :return True if the alarm is acknowledged; False otherwise
        '''
        return self==AlarmState.SET_ACK or self==AlarmState.CLEAR_ACK

    def ack(self):
        '''
        The alarm has been acknowledged (by the operator)
        :return:
        '''
        if self==AlarmState.SET_ACK or self==AlarmState.CLEAR_ACK:
            return self
        elif self==AlarmState.SET_UNACK:
            return AlarmState.SET_ACK
        else: # CLEAR_UNACK
            return AlarmState.CLEAR_ACK

    def set(self):
        '''
        The alarm has been set (by the TF)
        :return:
        '''
        if self==AlarmState.SET_ACK or self==AlarmState.SET_UNACK:
            return self
        else:
            return AlarmState.SET_UNACK

    def clear(self):
        if self==AlarmState.CLEAR_ACK or self==AlarmState.CLEAR_UNACK:
            return self
        elif self==AlarmState.SET_ACK:
            return AlarmState.CLEAR_ACK
        else: #SET_UNACK
            return AlarmState.CLEAR_UNACK

    @staticmethod
    def value_of(name):
        '''
        Get and return the AlarmState from the passed name.

        It delegates to Enum.__getitem__(cls, name)
        :param cls:  the class
        :param name: the name of the alarm state
        :return: the state with the given name
        :raise KeyError if the name does not correspond to a priority
        '''
        return AlarmState.__getitem__(name)

    def to_string(self):
        '''

        :param cls: the class
        :return: the string representation of the alarm state
        '''
        return self.name

    def __str__(self):
        return self.name
